### ISSUE TITLE
Chrome custom Profile upload issue

### ISSUE THREAD

**Issue by Anticope12** [aerokube/selenoid#1173]
Hello Team, I hope you are well, #1172  #1150

I've read several issues across Selenoid/Selenium and chromeDrivers without finding a solution. 

Currently, I'm facing an issue with adding a custom chrome profile, I've completed the following steps: 
1) volume setup.
2) using User-data-dir 
3) Profile directory. 
4)Broswer.json is up to   ##date.

and it's not working, with that being said, If I use the code on a local machine Using Selenoid it's working, I've checked Chrome: versions, and apparently. when the new sessions have created the default(User-data) being stored in a randomized generated directory.

Which it's makes it impossible to point a volume toward a specific Generated folder. 

I'm wondering if there is a way we could point it to pick up the custom profile for a specific path within the container using Volume. 

Is there a way to configure Selnoid-server to start a Chrome node with a given --user-data-dir value?  

Here is some information I've found about 
ChromeOptions options = new ChromeOptions();
options.addArguments("user-data-dir=/path/to/your/custom/profile");
You can create your own custom profile by just running Chrome (on the command-line or through
ChromeDriver) with the 'user-data-dir' switch set to some new directory. If the path doesn't exist, Chrome
will create a new profile in the specified location. You can then modify the profile settings as desired, and
ChromeDriver can use the profile in the future. Goto chrome://version in the browser to see what profile
Chrome is used.

![Screenshot 2022-01-30 at 03 28 27](https://user-images.githubusercontent.com/98564962/151685549-3c1b5835-b7d3-45b2-a420-b296d3c745b4.png)

On another note as well. when I try to access the custom profile from The session using (File:/// ), it gets stuck on loading.

I'm not sure what else, I could provide. 
Looking forward to hearing your thoughts about it.

**Comment by vania-pooh** [comment]
@Anticope12 so far as I know Chromedriver does not allow to upload a profile like Geckodriver does. However using capabilities (https://chromedriver.chromium.org/capabilities) you can upload extensions, change Chrome prefs, mobile emulation settings and so on separately. What is you use case?

**Comment by Anticope12** [comment]
Thank you for the clarification really appreciate it.
The Extention upload it's working without any issues( I did actually find the solution on one of the Q/A here).
 The reason I want to use the "profile/--profile-Directory", is just to save time going through all the Sign-in processes/ for each test, as well the session data/cache /cookies will be stored in the Custome profile directory which is using Volume "resistant data", therefore, it will allow me to use the same history, activity for a specific profile under new test/session.
 
 If you have an idea or workaround Would be great to hear.

**Comment by Anticope12** [comment]
Out of curiosity! I came across ChromeDriver Documentation: 
Should it be possible if ChromeDriver can create, As well read from a specific directory? 

As what I can see it's happening now is the following : 

"When chromedriver starts Chrome, if the ChromeOptions it receives from A"server" (which were passed as ChromeOptions to RemoteWebDriver) do not contain a value for --user-data-dir, then chromedriver generates a random chrome profile directory in /tmp and starts chrome with that value in --user-data-dir.  i.e. something like --user-data-dir=/tmp/.com.google.Chrome.ex4vLx/ 

Is there a way to configure A-server to start a Chrome node with a given --user-data-dir value?  "


 **Common use cases
Use custom profile (also called user data directory)**

By default, ChromeDriver **will create a new temporary profile** for **each session**. At times you may want to set special preferences or just use a **custom profile** altogether. If the former, you can use the 'chrome.prefs' capability (described later below) to specify preferences that will be applied after Chrome starts. If the latter, you can use the user-data-dir Chrome command-line **switch to tell Chrome which profiles to use**:

ChromeOptions options = new ChromeOptions();

options.addArguments("user-data-dir=/path/to/your/custom/profile");

You can create your own custom profile by just running Chrome (on the command-line or through ChromeDriver) with the user-data-dir switch set to some new directory. If the path doesn't exist, **Chrome will create a new profile** in the specified location. You can then modify the profile settings as desired, and ChromeDriver can use the profile in the future. Open chrome://version in the browser to see what profile Chrome is using.

**Comment by Anticope12** [comment]
Just adding some extra content here, 
I was playing around with the argument. "--user-data-dir" actually created a new path/directory Here is the screenshot, 

![Screenshot 2022-01-30 at 22 51 35](https://user-images.githubusercontent.com/98564962/151721606-9dbedf0f-b215-4d74-979f-3c44bc504c6d.png)

Here a snapshot if the code 
        options1 = webdriver.ChromeOptions()
   
        options1.add_argument('--user-data-dir=/tmp/chrome:')
        options1.add_argument('--profile-directory=Profile 1')

**Comment by vania-pooh** [comment]
@Anticope12 what you are describing will only work when Chromedriver is running on the same host where the test is being run. However in the majority of situations that's not the case. You need to first of all somehow transfer your profile to remote host with browsers and this is not implemented in Chromedriver.

**Comment by Anticope12** [comment]
Thank you @vania-pooh, for your feedback. Any thoughts of a workaround that could work here?

**Comment by Anticope12** [comment]
Is it normal to have a multiple user-date-dir created? 

I thought it was an issue on my end, I've created a brand new host and the system is creating multip DIR/Duplicate. 

![Screenshot 2022-01-31 at 14 23 39](https://user-images.githubusercontent.com/98564962/151810656-4bde93f4-7fa5-4f36-a5d5-e6327e376294.png)

**Comment by vania-pooh** [comment]
@Anticope12 yeah, we've also seen multiple directories like this in /tmp. Seems to be normal.

**Comment by aandryashin** [comment]
Hello, you can distribute custom chrome profiles over selenoid nodes and
use volumes to mount them to container, then pass user-data-dir option with
profile path.

вт, 1 февр. 2022 г., 07:15 Ivan Krutov ***@***.***>:

**Comment by Anticope12** [comment]
Hi @aandryashin , Thank you for the swift response. 
Do you mean to run something link the below:

1)store the Custom chrome profile Files on a docker.
2)Create a volume 
3)list the volume in Broswer.json
4)Execute the code below, If yes. I've already done it. and it's crashing with errors. For some reason, it cannot use the below Directory. (I've tried to change it, and had the same issue)  

options1 = webdriver.ChromeOptions()
 options1.add_argument('--user-data-dir=/tmp/chrome') # (If I use --user-data-dir=/tmp/chrome1 it does create a new Folider Chrome1 with a profile information within it.) 
 
        options1.add_argument('--profile-directory=Profile 1')
capabilities = dict(browserName="chrome", version="97.0", enableVNC=True, enableVideo=True,
                            volumes=["/var/lib/docker/volumes/myvolume/chrome:/tmp/chrome:ro"],
                            )

        self.driver = webdriver.Remote(
            command_executor="http://SelenoidIP:4444/wd/hub",
            desired_capabilities=capabilities,
            options=options1)
            
           
           
Please let me know if there is anything I could share that Could help you analyze the Issue. 
P.S: That's my first time using GitHub. 

THank you,

**Comment by aandryashin** [comment]
Hello, volumes are only supported in browsers.json and not in capabilities.

вт, 1 февр. 2022 г. в 15:14, Anticope12 ***@***.***>:

**Comment by Anticope12** [comment]
Hi @aandryashin , Thank you for the heads up, I was using only Volume without mounting it. 

For some reason, It's not loading any information from the profile for some reason, login process/bookmarks.  

Is there anything I could do/use to overcome this challenge? I've tried to use :ro to activate readonly. but it's not working either.

Thank you again for all of your support here.
Really appreciate it, It's been 5 days debugging process overall

**Comment by Anticope12** [comment]
Hi @aandryashin @vania-pooh , 
Sorry to bring this topic up, I'm just struggling to access the mount folder,
I'm using the following structure: 
        "volumes": [
    
                    "/var/www/new_directory/folder_profile:/1234/chrome"

        ],
         
I have the following info added in the browsers.json 
I've uploaded the Profile folder to the directory, However, I cannot access or open this folder when I run the sessions and go to file:///1234/chrome/ in the Chrome browser 

The page gets stuck in loading, I've tried other folders 
And this one has the same issue 
file:///root/

Would really really appreciate some guidance. 
P.S: Selenoid is no longer available in Digital Oscen marketplace

**Comment by vania-pooh** [comment]
@Anticope12 
1) We removed Selenoid from DigitalOcean because it was not really and delivering the same stuff with a Docker VM is a 5 minute task.
2) I would try to mount under `/home/selenium` because browsers using this user.

**Comment by Anticope12** [comment]
Thank you so much for your swift response @vania-pooh  for the heads up on 1. 

Regarding the 2nd part, I've tried your solution, to mount it under           "/var/www/new_directory/New:/home/selenium/test3:ro"
But unfortunately, the results are the same, I've also tried to mount it in different folders in the container, and got the same results. Anything that is mounted via volume to the container just gets sucked at the loading. using File://// in the browser.

Any thoughts on what could be the issue?

**Comment by vania-pooh** [comment]
@Anticope12 I would actually use `docker exec` to check whether files actually appear in container and whether you can read this files with regular Linux commands like `cat`, `less` and so on.

**Comment by Anticope12** [comment]
Hi @vania-pooh , Thank you for sending this over! apparently, the issue is related to permission issues, 
I've tried to access the container with root and I was able to access the folder and the files. 

I was trying to find a solution for it, but I was not able to do so, as the only option I've found online was to update the permission access for "ACTIVE Containers" only. 

I'm wondering if you have better options or if this is an option I've missed there. 
My experience with Linux and docker is still at a Rookie status. 


And thank you again for your swift replies and support 🙏
